### DISCUSSION TITLE
Documentation about EOF?

### DISCUSSION THREAD

**Discussion by scarf005** [lark-parser/lark#1175] [Q&A] [answered]
I'm trying to write a newline delimited language, but when i pass in single line program into parser it gives error since it's got EOF character, not newline character. so I tried to add EOF but it gave me this error.
```ml
code: expr ("\n" | $)
    | stmt
```
```py
  File "/home/scarf/Repo/etc/pico/.venv/lib/python3.10/site-packages/lark/load_grammar.py", line 964, in _parse_grammar
    raise GrammarError("Unexpected input at line %d column %d in %s: \n\n%s" %
lark.exceptions.GrammarError: Unexpected input at line 8 column 20 in <string>: 

code: expr ("\n" | $)
                   ^
```
i've read #237 #880 but despite their description, $ does not seems work. what should i do? my lark version is `1.1.2`.

**Comment by MegaIng** [comment, marked as answer]
#880 is unmergered, so that is why it doesn't work. You can manually clone that PR and see if it works in your local situation, but it has a few fundamental issues that makes it currently not something that can be merged easily.

**Reply by scarf005** [reply]
thanks for the answer. i'm still not sure how to handle newline as delimiter while using whitespace insignificant grammar (except \n)...
```
%import common.WS
%ignore WS
```

**Reply by erezsh** [reply]
The easiest solution is to always add `\n` at the end of your input, before you start parsing.
